File: 337.house-robber-iii.go
/*
 * @lc app=leetcode id=337 lang=golang
 *
 * [337] House Robber III
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func rob(root *TreeNode) int {
	v1, v2 := dfsRobTree(root)

	return max(v1, v2)
}

func dfsRobTree(root *TreeNode) (int, int) {
	if root == nil {
		return 0, 0
	}

	l0, l1 := dfsRobTree(root.Left)
	r0, r1 := dfsRobTree(root.Right)

	tmp0 := max(l0, l1) + max(r0, r1) // do not rob current node
	tmp1 := root.Val + l0 + r0        // rob current node

	return tmp0, tmp1
}

func max(a, b int) int {
	if a < b {
		return b
	}

	return a
}

// @lc code=end

File: 337.house-robber-iii.cpp
/*
 * @lc app=leetcode id=337 lang=cpp
 *
 * [337] House Robber III
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
 public:
  int rob(TreeNode* root) {
    auto r = rob_(root);
    return max(r.first, r.second);
  }

 private:
  // <first> rob cur
  // <second> not rob cur
  pair<int, int> rob_(TreeNode* node) {
    if (node == nullptr) return {0, 0};

    auto l = rob_(node->left);
    auto r = rob_(node->right);

    int val1 = node->val + l.second + r.second;
    int val2 = max(l.first, l.second) + max(r.first, r.second);

    return {val1, val2};
  }
};
// @lc code=end

